vmm/src/memory_manager.rs

```diff
@@ -1582,6 +1585,21 @@ impl MemoryManager {
 
         Ok(())
     }
+    #[cfg(all(feature = "mshv", target_arch = "x86_64"))]
+    //
+    // Live migration is compete and we want to set the dirty bits in the
+    // dirty log bitmap so that hypervisor can reconstruct the large pages
+    //
+    pub fn end_memory_dirty_log(&self) -> std::result::Result<(), MigratableError> {
+        for r in &self.guest_ram_mappings {
+            self.vm
+                .get_dirty_log(r.slot, r.gpa, r.size, DIRTY_BITMAP_SET_DIRTY)
```

Comment:
Can you elaborate more on what 's the behavior of using `DIRTY_BITMAP_SET_DIRTY` (comparing with `DIRTY_BITMAP_CLEAR_DIRTY`)? Also why is it needed?

Reply:
Note that: `guest_ram_mappings` only contains memory regions for guest RAM (a subset of all guest memory regions kvm/mshv created).

Is it intended that only this "subset" of the guest memory regions requires a call to mshv with `DIRTY_BITMAP_SET_DIRTY`? What about the rest memory regions?

Reply:
Set dirty tells the hypervisor that it either completes the migration or aborts so that hypervisor can rebuild the large pages afterward. While we enabled the tracking hypervisor breaks the large pages.

Reply:
I see. Then it sounded like you will want to do "set dirty" for all regions, as you enabled the tracking for all regions from `enable_dirty_page_tracking` (as you explained above).

Sounds like "set dirty" should be a part of `disable_dirty_page_tracking`. Any reason not do this as a part of `hypervisor::vm::disable_dirty_page_tracking()`?

Reply:
Sorry. One correction here. Hypervisor breaks the larges upon clearing the bits, not during the enable_tracking.
So disbale_tracking does not have any info on whether bits were cleared or not. But if you clear the birts you must set the bits before disabling the tracking.

Reply:
Similar to @rbradford's comments, I don't understand why we have to expose those `mshv` flags to the `vmm` crate, and they should be encapsulated at least in the `hypervisor` crate. Like what I asked earlier, why it can't be done as a part of `hypervisor::vm::disable_dirty_page_tracking()`?

Reply:
disable_dirty_page_tracking does not have any info to call get_dirty_log like gps, size etc.

Reply:
Well. That's why I proposed to add the following two methods to the `hypervisor::Vm` Trait in #2894. It aims to support both `kvm` and `mshv` while hiding hypervisor-specific information (say control flags for `kvm` and `mshv`:

```
    /// Start logging dirty pages
    fn start_dirty_log(
        &self,
        slot: u32,
        guest_phys_addr: u64,
        memory_size: u64,
        userspace_addr: u64,
    ) -> Result<()>;
    /// Stop logging dirty pages
    fn stop_dirty_log(
        &self,
        slot: u32,
        guest_phys_addr: u64,
        memory_size: u64,
        userspace_addr: u64,
    ) -> Result<()>;
```

I can understand you will need to expand the current `hypervisor::Vm::get_dirty_log()` method with taking `gpa` as input for `mshv`. Anything else is missing?

Reply:
I think this is enough.